Source/WebGPU/WGSL/CMakeLists.txt

```diff
@@ -0,0 +1,214 @@
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+# Build wgslc.
+# For simplicity, build it as a monolithic executable.
+
+# Target 'wgsl-types' generates TypeDeclarations.h in the build tree.
+add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/TypeDeclarations.h
+    COMMAND /usr/bin/env ruby
+        ${CMAKE_CURRENT_SOURCE_DIR}/generator/main.rb
+        ${CMAKE_CURRENT_SOURCE_DIR}/TypeDeclarations.rb
+        ${CMAKE_CURRENT_BINARY_DIR}/TypeDeclarations.h
+    DEPENDS
+        ${CMAKE_CURRENT_SOURCE_DIR}/generator/main.rb
+        ${CMAKE_CURRENT_SOURCE_DIR}/TypeDeclarations.rb
+    )
+add_custom_target(wgsl-types
+    ALL
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/TypeDeclarations.h)
+
+
+# Configure exectuable 'wgslc'
+
+set(wgslc_SOURCES
+    ./AttributeValidator.cpp
+    ./BoundsCheck.cpp
+    ./CallGraph.cpp
+    ./CompilationMessage.cpp
+    ./CompilationScope.cpp
+    ./ConstantValue.cpp
+    ./Constraints.cpp
+    ./EntryPointRewriter.cpp
+    ./GlobalSorting.cpp
+    ./GlobalVariableRewriter.cpp
+    ./Lexer.cpp
+    ./MangleNames.cpp
+    ./Overload.cpp
+    ./Parser.cpp
+    ./PointerRewriter.cpp
+    ./Token.cpp
+    ./TypeCheck.cpp
+    ./TypeStore.cpp
+    ./Types.cpp
+    ./VisibilityValidator.cpp
+    ./WGSL.cpp
+    ./WGSLEnums.cpp
+    ./WGSLShaderModule.cpp
+    ./wgslc.cpp
+
+    ./AST/ASTBinaryExpression.cpp
+    ./AST/ASTBuilder.cpp
+    ./AST/ASTDecrementIncrementStatement.cpp
+    ./AST/ASTStringDumper.cpp
+    ./AST/ASTUnaryExpression.cpp
+    ./AST/ASTVisitor.cpp
+
+    ./Metal/MetalCodeGenerator.cpp
+    ./Metal/MetalFunctionWriter.cpp
+    )
+
+set(wgslc_HEADERS
+    ./API.h
```

Comment:
Is the `./` prefix in file names really needed here? In theory CMake should already do the right thing and make relative paths depend on the current source directory.

Reply:
Heh, no.  I basically piped a `find` call.  Will remove them.

Reply:
fixed this.